src/TBuffer.cpp

```diff
@@ -2052,234 +2052,50 @@ void TBuffer::resetColors()
-            timeBuffer << blankTimeStamp;
-            promptBuffer << false;
-            firstChar = true;
-            continue;
-        }
-
-        // FIXME: (I18n) Need to measure painted line width and compare that
-        // to "unit" character width (whatever we work THAT out to be)
-        // multiplied by mWrap:
-        if (lineBuffer.back().size() >= mWrapAt) {
-            for (int i = lineBuffer.back().size() - 1; i >= 0; --i) {
-                // insert linebreak either at linebreaking character location or at last character of line
-                if (lineBreaks.indexOf(lineBuffer.back().at(i)) > -1 || i == 0) {
-                    const int linebreakPos = (i != 0) ? i + 1 : lineBuffer.back().size();
-                    QString tmp = lineBuffer.back().mid(0, linebreakPos);
-                    QString lineRest = lineBuffer.back().mid(linebreakPos);
-                    lineBuffer.back() = tmp;
-                    std::deque<TChar> newLine;
-
-                    int k = lineRest.size();
-                    if (k > 0) {
-                        while (k > 0) {
-                            newLine.push_front(buffer.back().back());
-                            buffer.back().pop_back();
-                            k--;
-                        }
-                    }
-
-                    buffer.push_back(newLine);
-                    if (lineRest.size() > 0) {
-                        lineBuffer.append(lineRest);
-                    } else {
-                        lineBuffer.append(QString());
-                    }
-                    timeBuffer << blankTimeStamp;
-                    promptBuffer << false;
-                    log(size() - 2, size() - 2);
-                    // Was absent causing loss of all but last line of wrapped
-                    // long lines of user input and some other console displayed
-                    // text from log file.
-                    break;
-                }
-            }
-        }
-        lineBuffer.back().append(text.at(i));
-        TChar c(fgColor, bgColor, (mEchoingText ? (TChar::Echo | flags) : flags), linkID);
-        buffer.back().push_back(c);
-        if (firstChar) {
-            timeBuffer.back() = QTime::currentTime().toString(timeStampFormat);
-            firstChar = false;
-        }
+    
+    int counter = 0;
+    for (int i = sub_start; i < sub_end; ++i) {
+        TChar format(fgColor, bgColor, (mEchoingText ? (TChar::Echo | flags) : flags), linkID);
+        buffer.back().push_back(format);
+        counter += 1;
+    }
+    QStringRef lineText = text.midRef(sub_start, sub_end - sub_start);
+    lineBuffer.back().append(lineText);
+    const QFont hostFont = mpHost->getDisplayFont();
+    int mFontWidth = QFontMetrics(hostFont).averageCharWidth();
+    int wrapByPixel = mWrapAt * mFontWidth;
+    QFontMetrics qfm(hostFont);
+    if (qfm.horizontalAdvance(lineBuffer.back()) > wrapByPixel || lineText.indexOf("\n") != -1) {
```

Comment:
```suggestion
    if (qfm.horizontalAdvance(lineBuffer.back()) > wrapByPixel || lineText.indexOf(QChar::LineFeed) > -1) {
```

Maybe this is faster/more intuitive?